python/lsst/daf/butler/core/mappingFactory.py

```diff
@@ -0,0 +1,144 @@
+#
+# LSST Data Management System
+#
+# Copyright 2008-2018  AURA/LSST.
+#
+# This product includes software developed by the
+# LSST Project (http://www.lsst.org/).
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
+# see <https://www.lsstcorp.org/LegalNotices/>.
+#
+from lsst.daf.butler.core.utils import doImport
+
+
+class MappingFactory:
+    """
+    Register the mapping of some key to a python type and retrieve instances.
+
+    Enables instances of these classes to be retrieved from the factory later.
+    The class can be specified as an object, class or string.
+    If the key is an object it is converted to a string by accessing
+    a `name` attribute.
+    """
+
+    refType = None
+    """Type of instances expected to be returned from factory."""
+
+    def __init__(self):
+        self._registry = {}
+
+    def getFromRegistry(self, targetClass, override=None):
+        """Get a new instance of the object stored in the registry.
+
+        Parameters
+        ----------
+        targetClass : `str` or object supporting `.name` attribute
+            Get item from registry associated with this target class, unless
+        override : `str` or object supporting `.name` attribute, optional
+            If given, look if an override has been specified for this and,
+            if so return that instead.
```

Comment:
Perhaps make it slightly more generic? By taking `*targetClasses` and thus allowing for multiple overrides?